qadence/mitigations/readout.py

```diff
@@ -0,0 +1,94 @@
+from __future__ import annotations
+
+from collections import Counter
+from functools import reduce
+
+import numpy as np
+import numpy.typing as npt
+import torch
+from scipy.linalg import norm
+from scipy.optimize import LinearConstraint, minimize
+
+from qadence.mitigations.protocols import Mitigations
+from qadence.noise.protocols import Noise
+
+
+def corrected_probas(p_corr: npt.NDArray, T: npt.NDArray, p_raw: npt.NDArray) -> np.double:
+    return norm(T @ p_corr.T - p_raw.T, ord=2) ** 2
+
+
+def renormalize_counts(corrected_counts: npt.NDArray, n_shots: int) -> npt.NDArray:
+    """Renormalize counts rounding discrepancies."""
+    total_counts = sum(corrected_counts)
+    if total_counts != n_shots:
+        counts_diff = total_counts - n_shots
+        corrected_counts -= counts_diff
+        corrected_counts = np.where(corrected_counts < 0, 0, corrected_counts)
+        sum_corrected_counts = sum(corrected_counts)
+        if sum_corrected_counts < n_shots:
+            renormalization_factor = max(sum_corrected_counts, n_shots) / min(
+                sum_corrected_counts, n_shots
+            )
+        else:
+            renormalization_factor = min(sum_corrected_counts, n_shots) / max(
+                sum_corrected_counts, n_shots
+            )
+        corrected_counts = np.rint(corrected_counts * renormalization_factor).astype(int)
+    return corrected_counts
+
+
+def mitigation_minimization(
+    noise: Noise, mitigation: Mitigations, samples: list[Counter]
+) -> list[Counter]:
+    """Minimize a correction matrix subjected to stochasticity constraints.
+
+    See Equation (5) in https://arxiv.org/pdf/2001.09980.pdf.
+    """
+    corrected_counters: list[Counter] = []
+    for sample in samples:
+        n_qubits = len(list(sample.keys())[0])
+        n_shots = sum(sample.values())
+        bitstring_length = 2**n_qubits
+        noise_matrices = noise.options.get("noise_matrix", noise.options["confusion_matrices"])
+        # Build the whole T matrix.
+        T_matrix = reduce(torch.kron, noise_matrices).detach().numpy()
+        # List of bitstrings in lexicographical order.
+        ordered_bitstrings = [f"{i:0{n_qubits}b}" for i in range(bitstring_length)]
+        # Tensor of raw probabilites.
+        p_raw = torch.tensor([sample[bs] for bs in ordered_bitstrings]) / n_shots
```

Comment:
Can p_raw directly be  np.array([sample[bs] for bs in ordered_bitstrings]) / n_shots or the tensor is needed for some precision aspect?